mutation_writer/token_group_based_splitting_writer.cc

```diff
@@ -45,8 +45,8 @@ class token_group_based_splitting_mutation_writer {
 
         if (_current_writer && _current_group_id > prev_group_id) [[unlikely]] {
             _current_writer->consume_end_of_stream();
-            return _current_writer->close().then([this] {
-                _current_writer = std::nullopt;
+            bucket_writer_v2 wr = std::move(*std::exchange(_current_writer, std::nullopt));
+            return wr.close().then([this, wr = std::move(wr)] {
                 allocate_new_writer_if_needed();
```

Comment:
makes sense.makes sense. please confirm my understanding.

1) current writer is being closed.
2) split is stopped
3) the consumer-producer loop catches the exception, and closes the mutation splitting writer
4) current writer is asked to close again, but bucket_writer_v2::close() doesn't allow for it

looks like we cannot prevent the loop from calling close() while current writer is being closed, yet token_group_based_splitting_mutation_writer::close() called on step 4, should only return once close() initiated on step 1 returns, to avoid use-after-free.

Reply:
Not sure about the outer layer, but at the low level what I think happened is rhat once the writer is aborted, _consume_fut would return an error, and once maybe_switch_to_new_writer returns that error from the call to close(), the .then continuation doesn't run and _current_writer isn't reset, and so close() tries to close it again and hits use after move in bucket_writer_v2::close().

Reply:
I think that in addition to that we might face a use after free if writer is aborted during ongoing close when switching writer since close() in the loop will find a ready future thinking current writer is closed already. makes sense?

Reply:
I took a closer look, and I don't think what I described above could happen